src/foam/dao/invoiceHistoryAuthorizerTest.js

```diff
@@ -0,0 +1,145 @@
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'invoiceHistoryAuthorizerTest',
+  extends: 'foam.nanos.test.Test',
+  flags: ['java'],
+
+  javaImports: [
+    'foam.core.X',
+    'foam.dao.DAO',
+    'foam.dao.SequenceNumberDAO',
+    'foam.nanos.auth.User',
+    'foam.nanos.test.Test',
+    'foam.test.TestUtils',
+    'static foam.mlang.MLang.*',
+    'net.nanopay.invoice.model.Invoice',
+    'foam.dao.history.HistoryRecord',
+    'foam.dao.ArraySink',
+    'net.nanopay.contacts.Contact',
+    'net.nanopay.auth.PublicUserInfo',
+    'foam.nanos.auth.AuthorizationException',
+    'java.util.Date'
+  ],
+
+  documentation: 'Class to test invoiceHistoryDAO security',
+
+  methods: [
+    {
+      name: 'runTest',
+      javaCode: `
+        x = x.put("user", null)
+             .put("group", null)
+             .put("twoFactorSuccess", false);
+        User payerUser = new User();
+        User payeeUser = new User();
+        User unrelatedUser = new User();
+        User adminUser = new User();
+        DAO invoiceHistoryDAO = ((DAO) x.get("invoiceHistoryDAO")).inX(x);
+        DAO invoiceDAO = (DAO) x.get("invoiceDAO");
+        DAO bareUserDAO = (DAO) x.get("bareUserDAO");
+        Invoice invoice = new Invoice();
+        HistoryRecord historyRecord = new HistoryRecord();
+        boolean threw;
+
+        payerUser.setId(1);
+        payeeUser.setId(2);
+        unrelatedUser.setId(3);
+        adminUser.setGroup("admin");
+
+        x = x.put("user", adminUser);
+
+        payeeUser = (User) bareUserDAO.put(payeeUser);
+
+        invoice.setPayeeId(payeeUser.getId());
+        invoice.setPayerId(payerUser.getId());
+        invoice = (Invoice) invoiceDAO.put(invoice);
+
+        historyRecord.setObjectId(invoice.getId());
+        historyRecord = (HistoryRecord) invoiceHistoryDAO.inX(x).put(historyRecord);
+
+        x = x.put("user", payerUser);
+
+        ArraySink invoiceHistoryTestSinc = (ArraySink) invoiceHistoryDAO.inX(x).where(
+          EQ(HistoryRecord.OBJECT_ID, historyRecord.getObjectId())
+        ).select(new ArraySink());
+
+        test( ! (invoiceHistoryTestSinc.getArray().size() == 0) , "User (Payer) from the same buisness can view invoice");
+
+        x = x.put("user", payeeUser);
+        invoiceHistoryTestSinc = (ArraySink) invoiceHistoryDAO.inX(x).where(
+          EQ(HistoryRecord.OBJECT_ID, historyRecord.getObjectId())
+        ).select(new ArraySink());
+
+        test( ! (invoiceHistoryTestSinc.getArray().size() == 0) , "User (Payee) from the same buisness can view invoice");
+
+        x = x.put("user", unrelatedUser);
+        invoiceHistoryTestSinc = (ArraySink) invoiceHistoryDAO.inX(x).where(
+          EQ(HistoryRecord.OBJECT_ID, historyRecord.getObjectId())
+        ).select(new ArraySink());
+
+        test( invoiceHistoryTestSinc.getArray().size() == 0 , "User from different buisness can not view invoice");
+
+        threw = true;
```

Comment:
This is backwards. It hasn't thrown yet, so this should be false by default and should be set to true in the catch block. Same elsewhere in this file.